libopenage/gui/minimap.cpp

```diff
@@ -0,0 +1,397 @@
+			auto big_side = std::max(this->boundingRect().width(), this->boundingRect().height());
+			auto small_side = std::min(this->boundingRect().width(), this->boundingRect().height());
+
+			auto background_gui_tex = gui_texture_from_img_data(std::move(std::get<0>(bg_update.background_tex)), std::get<1>(bg_update.background_tex), std::get<2>(bg_update.background_tex));
+
+			background->setTexture(std::get<std::unique_ptr<GuiTexture>>(background_gui_tex).release());
+			background->setSourceRect(std::get<QRectF>(background_gui_tex));
+
+			auto terrain_bounding_square_max_side_len = terrain_square_dimensions(bg_update.bounding_square);
+			coord::phys_t max_terrain_side_len = coord::chunk{terrain_bounding_square_max_side_len, 0}.to_tile({0, 0}).to_phys2({0, 0}).as_relative().ne;
+			QMatrix4x4 downscale_world;
+			downscale_world.translate(big_side / 2, big_side / 2);
+			downscale_world.scale(big_side / max_terrain_side_len, big_side / max_terrain_side_len);
+			auto terrain_bounding_rect_center_phys = terrain_square_center(bg_update.bounding_square).to_tile3().to_phys3();
+			downscale_world.translate(-terrain_bounding_rect_center_phys.ne, -terrain_bounding_rect_center_phys.se);
+			world_to_background->setMatrix(downscale_world);
+
+			if (this->last_bounding_rect != this->boundingRect()) {
+				QMatrix4x4 trans_to_center(QTransform::fromTranslate(-big_side / 2., -big_side / 2.));
+
+				auto iv = (coord::tile3{1, 0, 0}.to_phys3() - coord::tile3{0, 0, 0}.to_phys3()).to_camgame().to_window();
+				auto jv = (coord::tile3{0, 1, 0}.to_phys3() - coord::tile3{0, 0, 0}.to_phys3()).to_camgame().to_window();
+
+				/**
+				 * Background texture uses square pixels for tiles, so need to undo the tile sizing.
+				 */
+				auto i = QVector2D(iv.x / engine_coord_data.tile_halfsize.x, iv.y / engine_coord_data.tile_halfsize.y);
+				auto j = QVector2D(jv.x / engine_coord_data.tile_halfsize.x, jv.y / engine_coord_data.tile_halfsize.y);
+
+				QMatrix4x4 cam_rotate(
+					i.x(), j.x(), 0.f, 0.f,
+					i.y(), j.y(), 0.f, 0.f,
+					0.f, 0.f, 1.f, 0.f,
+					0.f, 0.f, 0.f, 1.f
+				);
+
+				/**
+				 * Fit a rotated square into the same area.
+				 */
+				auto diagonal = cam_rotate * QVector4D(1.f, 1.f, 0.f, 0.f);
+				auto diagonal_max_coord = std::max(diagonal.x(), diagonal.y());
+				QMatrix4x4 scale_fit(QTransform::fromScale(1. / diagonal_max_coord, 1. / diagonal_max_coord));
+
+				auto ratio = width_bigger ? std::make_tuple(1., small_side / big_side) : std::make_tuple(small_side / big_side, 1.);
+				QMatrix4x4 aspect(QTransform::fromScale(std::get<0>(ratio), std::get<1>(ratio)));
+
+				QMatrix4x4 trans_from_center(QTransform::fromTranslate(this->boundingRect().width() / 2., this->boundingRect().height() / 2.));
+
+				turn->setMatrix(trans_from_center * aspect * scale_fit * cam_rotate * trans_to_center);
+
+				background->setRect(0., 0., big_side, big_side);
+
+				this->last_bounding_rect = this->boundingRect();
```

Comment:
the draw methods could use some explainations how the minimap creation/drawing works.